src/memray/_memray/record_writer.cpp

```diff
@@ -93,23 +392,266 @@ RecordWriter::writeTrailer()
     return writeSimpleType(token);
 }
 
-std::unique_lock<std::mutex>
-RecordWriter::acquireLock()
+std::unique_ptr<RecordWriter>
+StreamingRecordWriter::cloneInChildProcess()
 {
-    return std::unique_lock<std::mutex>(d_mutex);
+    std::unique_ptr<io::Sink> new_sink = d_sink->cloneInChildProcess();
+    if (!new_sink) {
+        return {};
+    }
+    return std::make_unique<StreamingRecordWriter>(
+            std::move(new_sink),
+            d_header.command_line,
+            d_header.native_traces);
+}
+
+AggregatingRecordWriter::AggregatingRecordWriter(
+        std::unique_ptr<memray::io::Sink> sink,
+        const std::string& command_line,
+        bool native_traces)
+: RecordWriter(std::move(sink))
+{
+    memcpy(d_header.magic, MAGIC, sizeof(d_header.magic));
+    d_header.version = CURRENT_HEADER_VERSION;
+    d_header.native_traces = native_traces;
+    d_header.file_format = FileFormat::AGGREGATED_ALLOCATIONS;
+    d_header.command_line = command_line;
+    d_header.pid = ::getpid();
+    d_header.python_allocator = getPythonAllocator();
+
+    d_stats.start_time = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+}
+
+void
+AggregatingRecordWriter::setMainTidAndSkippedFrames(
+        thread_id_t main_tid,
+        size_t skipped_frames_on_main_tid)
+{
+    d_header.main_tid = main_tid;
+    d_header.skipped_frames_on_main_tid = skipped_frames_on_main_tid;
+}
+
+bool
+AggregatingRecordWriter::writeHeader(bool seek_to_start)
+{
+    // Nothing to do; everything is written by writeTrailer.
+    (void)seek_to_start;
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeTrailer()
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+
+    d_stats.end_time = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+    d_header.stats = d_stats;
+    if (!writeHeaderCommon(d_header)) {
+        return false;
+    }
+
+    for (const auto& memory_snapshot : d_memory_snapshots) {
+        if (!writeSimpleType(AggregatedRecordType::MEMORY_SNAPSHOT) || !writeSimpleType(memory_snapshot))
+        {
+            return false;
+        }
+    }
+
+    for (const auto& [tid, thread_name] : d_thread_name_by_tid) {
+        if (!writeSimpleType(AggregatedRecordType::CONTEXT_SWITCH)
+            || !writeSimpleType(ContextSwitch{tid})
```

Comment:
Why do we need to simulate a context switch here?

Reply:
Just to allow us to use the same format for `THREAD_RECORD` between the aggregated capture file format and the all-allocations capture file format. The old file format didn't include the tid in the thread record, and instead the thread record was implicitly related to the last thread switched to. I matched that here only to allow reusing code in the reader.